### PR TITLE
Terminate the instance when 404 occured.

### PR DESCRIPTION
**Pull request by Artmorse**
This PR is based on #470.

This wants to fix the #469 issue.

### Manual testing

I've configured a GCE cloud with the specific label `gcp-ce`.
I've created a dummy `hello-world` pipeline using GCE nodes (using the label `gcp-ce`).
I've provisionned a node from the cloud.
Once the compute engine up and running, I ran the `hello-world` pipeline and it working as expected.
Then I've deleted it using the GCP UI the node is well terminated and removed from Jenkins.
I re-ran the pipeline and a new node has been provisionned.

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [X] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineComputerLauncher.java [hunk 1]
```diff
@@ -410,11 +422,25 @@ protected Connection connectToSsh(ComputeEngineComputer computer, TaskListener l
...
                 // keep retrying until SSH comes up
-                logInfo(computer, listener, "Failed to connect via ssh: " + e.getMessage());
-                logInfo(computer, listener, "Waiting for SSH to come up. Sleeping 5.");
+                logInfo(computer, listener, String.format("Failed to connect via ssh: %s", e.getMessage()));
+                logInfo(
+                        computer,
+                        listener,
+                        String.format("Waiting for SSH to come up. Sleeping %d.", SSH_SLEEP_MILLIS / 1000));
                 Thread.sleep(SSH_SLEEP_MILLIS);
+            } catch (IOException e) {
```
Comment (dwnusbaum) [inline src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineComputerLauncher.java:445]: Should we preserve the `sleep` in this case to make sure we don't retry in a tight loop with no sleep? Or did you want to terminate the node here?

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineInstance.java [hunk 2]
```diff
@@ -130,9 +134,16 @@ protected void _terminate(TaskListener listener) throws IOException, Interrupted
...
-            // If the instance is running, attempt to terminate it. This is an async call and we
+            Map<String, String> filterLabel = ImmutableMap.of(CLOUD_ID_LABEL_KEY, cloud.getInstanceId());
+            var instanceExistsInCloud =
+                    cloud.getClient().listInstancesWithLabel(cloud.getProjectId(), filterLabel).stream()
+                            .anyMatch(instance -> instance.getName().equals(name));
+
+            // If the instance exists in the cloud, attempt to terminate it. This is an async call and we
             // return immediately, hoping for the best.
-            cloud.getClient().terminateInstanceAsync(cloud.getProjectId(), zone, name);
+            if (instanceExistsInCloud) {
```
Comment (dwnusbaum) [inline src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineInstance.java:144]: Why is this needed? Is it just to avoid an unnecessary error message in the logs in case the instance does not exist in the cloud?

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineComputerLauncher.java (hunk 1 above)
Comment (Artmorse) [reply src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineComputerLauncher.java:445]: I'm not sure if we want to terminate the node in this case. To keep in the same state as before, I will re-add the sleep when an IOException is thrown.

done into 627fda8ab8fd40a57386ac738ae4f759f1ff1af1

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineInstance.java (hunk 2 above)
Comment (Artmorse) [reply src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineInstance.java:144]: This PR is based on another branch (see #470) but I don't think this condition is useful. Theoretically we could even verify the instance exists in the cloud but when running the `terminateInstanceAsync` the instance could have been removed. I will remove that part.

Please @Theoderich if we're missing cases, please tell us.

Comment (Artmorse) [reply src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineInstance.java:144]: done into f6fc88ceb54d90dd77459fc4c5e20141aecc3858

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineComputerLauncher.java [hunk 3]
```diff
@@ -214,19 +218,23 @@ public void launch(SlaveComputer slaveComputer, TaskListener listener) {
...
-            }
+            terminateNode(slaveComputer, listener);
         } catch (InterruptedException ie) {
 
         }
     }
 
+    private static void terminateNode(SlaveComputer slaveComputer, TaskListener listener) {
+        ComputeEngineInstance node = (ComputeEngineInstance) slaveComputer.getNode();
+        if (node != null) {
```
Comment (jgreffe) [inline src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineComputerLauncher.java:229]: Should we add some fine logs if `node == null`, something like `"Tried to terminate unknown node from computer %s"` ?

**Review by jgreffe — COMMENTED** [review]
Are there some effective unit tests for this?

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineInstance.java (hunk 2 above)
Comment (Theoderich) [reply src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineInstance.java:144]: This is part of the fix for #471 and should not be removed.
**The Problem:**
When Jenkins tries to start a node and the quotas in the target cloud do not allow it, the instance is not started in the cloud but Jenkins thinks that a node was started. Now there is a Zombie node in Jenkins.

When trying to delete this Zombie node, Jenkins calls `cloud.getClient().terminateInstanceAsync` which throws an Exception since the Instance does not exist in the cloud. The Delete is aborted and the zombie node stays, forever undeletable.

**The Solution:**
I added a check if the node actually exists in the cloud before calling `cloud.getClient().terminateInstanceAsync`. If the node does not exist as an instance in the cloud, there is no need  to terminate it and the deletion can proceed.

**Why not just catch the exception thrown by `cloud.getClient().terminateInstanceAsync`?**
It is not clear from the exception why the terminate operation failed. If the instance exists in the cloud and the termination failed for some other reason, we do not want to delete the node in Jenkins.

**Comment by Artmorse** [comment]
> Are there some effective unit tests for this?

I don't think we can write any unit test for this (apart from maybe using Wiremock but this seems huge...).
This repository has some [integration tests](https://github.com/jenkinsci/google-compute-engine-plugin/blob/develop/CONTRIBUTING.md#integration-tests) but I failed to run them.

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineComputerLauncher.java (hunk 3 above)
Comment (Artmorse) [reply src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineComputerLauncher.java:229]: Mmmmh yes, we could.

Comment (Artmorse) [reply src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineComputerLauncher.java:229]: done into ce2791c144537bd34b83b04c159b5e7cb471c9f3

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineInstance.java (hunk 2 above)
Comment (Artmorse) [reply src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineInstance.java:144]: Ok, thanks for the detailed explanation.
I've reverted the changes.

**Comment by batmat** [comment]
@Artmorse can you please fix the build failure? Seems like SpotBugs issues?

I just labeled the issue as `bug`. As one of the [maintainers](https://github.com/jenkins-infra/repository-permissions-updater/blob/4bedc365625a5da948d8e974f7445a13e2b884b9/permissions/plugin-google-compute-engine.yml#L16), I will merge this once the build goes green.

**Comment by batmat** [comment]
@Artmorse merged. CD is enabled, so a release should be cut automatically within the next minutes.
